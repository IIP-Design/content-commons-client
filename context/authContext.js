import React, { useState } from 'react';
import { useQuery, useMutation, useApolloClient } from '@apollo/react-hooks';
import {
  CLOUDFLARE_SIGNIN_MUTATION,
  USER_SIGN_OUT_MUTATION,
  CURRENT_USER_QUERY
} from 'lib/graphql/queries/user';
import { useRouter } from 'next/router';
import { redirectTo, isDevEnvironment } from 'lib/browser';
import { isRestrictedPage } from 'lib/authentication';
import cookie from 'js-cookie';
import cookies from 'next-cookies';

const AuthContext = React.createContext();
const allowedRolesForRestrictedPages = ['EDITOR', 'TEAM_ADMIN', 'ADMIN'];

/**
 * Check user page permissions
 * Exported to make avaiale for SSR
 * @param {object} user graphQL user
 */
export const hasPagePermissions = user => user.permissions.some( permission => allowedRolesForRestrictedPages.includes( permission ) );

/**
 * Return the cookie that CloudFlare sets upon successful sign in to
 * @param {*} ctx next.js context object
 */
const getCloudFlareToken = ctx => {
  // CloudFlare is not connected to the dev environment
  // so send replacement
  if ( isDevEnvironment() ) {
    return 'cfToken';
  }

  /* eslint-disable camelcase */
  const { CF_Authorization } = cookies( ctx );
  return CF_Authorization;
  /* eslint-enable */
};

/**
 * Returns a user if both qa CloudFlare token AND
 * valid user exist, else return null
 * Exported to make avaiale for SSR
 * @param {obj} ctx next.js context object
 */
export const fetchUser = async ctx => {
  const cfAuth = getCloudFlareToken( ctx );

  // CloudFlare token MUST be available, if not return null
  if ( !cfAuth ) {
    return null;
  }

  try {
    const {
      data: { user }
    } = await ctx.apolloClient.query( { query: CURRENT_USER_QUERY, fetchPolicy: 'network-only' } );
    if ( user ) {
      // add token to authenticate to elastic api to user
      const { ES_TOKEN } = cookies( ctx );
      return { ...user, esToken: ES_TOKEN };
    }
    return null;
  } catch ( err ) {
    return null;
  }
};

/**
 * Checks to see if on a protected page and if so
 * verify a logged user with applicable permissions
 * Exported to make avaiale for SSR
 * @param {*} ctx ext.js context object
 */
export const canAccessPage = async ctx => {
  // are we on a page that requires permissions?
  if ( isRestrictedPage( ctx.pathname ) ) {
    const user = await fetchUser( ctx );

    // ensure a user and a user with proper permissions
    if ( !user || !hasPagePermissions( user ) ) {
      return false;
    }
  }
  return true;
};


function AuthProvider( props ) {
  const client = useApolloClient();
  const router = useRouter();

  // ensure signIn only called once
  const [attemptToSignIntoCF, setAttemptToSignIntoCF] = useState( false );


  // Attempt to fetch user
  const { data, loading: userLoading } = useQuery( CURRENT_USER_QUERY, {
    ssr: false,
    fetchPolicy: 'network-only'
  } );


  // Sign in mutation
  const [signIn, { loading: signInLoading, error: signInError }] = useMutation(
    CLOUDFLARE_SIGNIN_MUTATION, {
      refetchQueries: [{ query: CURRENT_USER_QUERY }]
    }
  );

  // Sign out mutation
  const [signOut] = useMutation( USER_SIGN_OUT_MUTATION, {
    onCompleted: () => {
      setAttemptToSignIntoCF( false );
      client.resetStore();

      // sign out of CF
      const {
        location: { protocol, hostname, port }
      } = window;
      const url = `${protocol}//${hostname}:${port}`;
      window.location = `https://america.cloudflareaccess.com/cdn-cgi/access/logout?returnTo=${url}`;
      cookie.remove( 'CF_Authorization' );
      cookie.remove( 'ES_TOKEN' );
      router.push( '/' );
    }
  } );


  const login = async () => {
    // do we have a CloudFlare token?
    const cfAuth = cookie.get( 'CF_Authorization' );
    if ( cfAuth ) {
      // ensure signIn only called once
      if ( !attemptToSignIntoCF ) {
        signIn( { variables: { token: cfAuth } } ).catch( err => console.dir( signInError ) );
        setAttemptToSignIntoCF( true );
      }

      if ( signInLoading ) {
        // console.log( 'Signing in' );
      }

      if ( signInError ) {
        // console.log( 'There was an error' )
      }

      redirectTo( '/', {} );
    }
  };


  const logout = async () => signOut();
  const register = () => console.log( 'register' );

  return (
    <AuthContext.Provider
      value={ {
        user: data?.user,
        loading: userLoading,
        login,
        logout,
        register
      } }
      { ...props }
    />
  );
}


function useAuth() {
  const context = React.useContext( AuthContext );
  if ( context === undefined ) {
    throw new Error( `useAuth must be used within a AuthProvider` );
  }
  return context;
}

export { AuthProvider, useAuth };
