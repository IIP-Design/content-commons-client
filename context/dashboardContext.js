import React from 'react';

import { normalizeDashboardData } from 'lib/graphql/util';
import { TEAM_GRAPHIC_PROJECTS_QUERY, TEAM_GRAPHIC_PROJECTS_COUNT_QUERY } from 'lib/graphql/queries/graphic';
import { TEAM_VIDEO_PROJECTS_QUERY, TEAM_VIDEO_PROJECTS_COUNT_QUERY } from 'lib/graphql/queries/video';
import { TEAM_PACKAGES_QUERY, TEAM_PACKAGES_COUNT_QUERY } from 'lib/graphql/queries/package';

const initialState = {
  content: {},
  count: {},
  queries: null,
  team: { contentTypes: null }
};

export const DashboardContext = React.createContext( initialState );

const testContentTypes = team => {
  const contentTypes = team?.contentTypes ? team.contentTypes : '';

  if ( contentTypes.includes( 'GRAPHIC' ) ) return 'graphic';
  if ( contentTypes.includes( 'VIDEO' ) ) return 'video';
  if ( contentTypes.includes( 'PACKAGE' ) ) return 'packages';

  return null;
};

const parseCount = ( queryData, team ) => {
  const type = testContentTypes( team );

  const count = queryData?.[type] ? queryData[type].length : 0;

  return count;
};

export const setQueries = team => {
  const queries = {};

  switch ( testContentTypes( team ) ) {
  case 'graphic':
    queries.content = TEAM_GRAPHIC_PROJECTS_QUERY;
    queries.count = TEAM_GRAPHIC_PROJECTS_COUNT_QUERY;

    return queries;
  case 'video':
    queries.content = TEAM_VIDEO_PROJECTS_QUERY;
    queries.count = TEAM_VIDEO_PROJECTS_COUNT_QUERY;

    return queries;
  case 'packages':
    queries.content = TEAM_PACKAGES_QUERY;
    queries.count = TEAM_PACKAGES_COUNT_QUERY;

    return queries;
  default:
    return queries;
  }
};

export const dashboardReducer = ( state, action ) => {
  switch ( action.type ) {
  case 'UPDATE_COUNT':
    return {
      ...state,
      count: {
        count: parseCount( action.payload.count.data, action.payload.team ),
        error: action.payload.count.error,
        loading: action.payload.count.loading,
        refetch: action.payload.count.refetch
      }
    };
  case 'UPDATE_CONTENT':
    return {
      ...state,
      content: {
        data: normalizeDashboardData( action.payload.data, action.payload.type ),
        error: action.payload.error,
        loading: action.payload.loading,
        refetch: action.payload.refetch
      }
    };
  case 'UPDATE_TEAM':
    return {
      ...state,
      team: action.payload.team,
      projectType: testContentTypes( action.payload.team ),
      queries: setQueries( action.payload.team )
    };
  default:
    return { ...state };
  }
};
