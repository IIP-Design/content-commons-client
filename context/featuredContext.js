import React from 'react';

const initialState = {
  priorities: {},
  recents: {},
  loading: false,
  lastLoad: 0,
  error: false,
};

export const FeaturedContext = React.createContext( initialState );

export const featuredReducer = ( state, action ) => {
  const { payload } = action;

  switch ( action.type ) {
    case 'LOAD_FEATURED_FAILED':
      return {
        ...state,
        loading: false,
        error: true,
      };
    case 'LOAD_FEATURED_PENDING':
      return {
        ...state,
        loading: true,
        error: false,
      };
    case 'LOAD_FEATURED_SUCCESS':
      return {
        ...state,
        priorities: payload.priorities,
        recents: payload.recents,
        loading: false,
        error: false,
        lastLoad: Date.now(),
      };
    default:
      return state;
  }
};
