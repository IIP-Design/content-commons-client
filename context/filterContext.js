import React from 'react';

const initialState = {
  categories: [],
  countries: [],
  sources: [],
  postTypes: [],
  dateFrom: new Date(),
  dateTo: new Date(),
  date: 'recent',
};

export const FilterContext = React.createContext( initialState );

const handlePayloadType = ( payload, state, prop ) => {
  // If there is no payload, clear selected property
  if ( !payload ) {
    return { ...state, [prop]: [] };
  }

  if ( typeof payload === 'string' ) {
    return { ...state, [prop]: [payload] };
  }

  return { ...state, [prop]: [...payload] };
};

export const filterReducer = ( state, action ) => {
  const { payload } = action;

  switch ( action.type ) {
    case 'CLEAR_FILTERS':
      return initialState;

    case 'FILTER_CATEGORY_CHANGE':
      return handlePayloadType( payload, state, 'categories' );

    case 'FILTER_COUNTRY_CHANGE':
      return handlePayloadType( payload, state, 'countries' );

    case 'FILTER_POST_TYPE_CHANGE':
      return handlePayloadType( payload, state, 'postTypes' );

    case 'FILTER_SOURCE_CHANGE':
      return handlePayloadType( payload, state, 'sources' );

    case 'FILTER_DATE_CHANGE':
      return {
        ...state,
        date: payload || 'recent',
      };

    case 'FILTER_FROM_DATE_CHANGE':
      return {
        ...state,
        dateFrom: payload,
      };

    case 'FILTER_TO_DATE_CHANGE':
      return {
        ...state,
        dateTo: payload,
      };

    default:
      return state;
  }
};
