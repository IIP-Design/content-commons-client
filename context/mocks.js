export const graphicMock = {
  graphicProjects: [
    {
      id: '1234',
      createdAt: '2019-05-09T18:33:03.368Z',
      updatedAt: '2019-05-09T18:33:03.368Z',
      publishedAt: '2019-05-09T18:33:03.368Z',
      type: 'LANGUAGE',
      title: 'Test title language',
      copyright: '',
      alt: 'Alternative text',
      descPublic:
               'It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout.',
      descInternal:
               'It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout.',
      assetPath:
               'https://cdp-video-tst.s3.amazonaws.com/2020/01/commons.america.gov_ck5iguwcv01j40729bz48sijs/coffee.jpg',
      author: {
        id: 'ndfkjbewkj',
        firstName: 'Jane',
        lastName: 'Doe',
        email: 'doej@america.gov',
      },
      team: {
        id: 'wjbjb',
        name: 'GPA Design & Editorial',
        contentTypes: ['GRAPHIC'],
      },
      status: 'PUBLISHED',
      visibility: 'PUBLIC',
      images: [],
      categories: [],
      tags: [],
      __typename: 'GraphicProject',
    },
    {
      id: '5678',
      createdAt: '2019-05-09T18:33:03.368Z',
      updatedAt: '2019-05-09T18:33:03.368Z',
      publishedAt: '2019-05-09T18:33:03.368Z',
      type: 'SOCIAL_MEDIA',
      title: 'Test title social media',
      copyright: '',
      alt: 'Alternative text',
      descPublic:
               'It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout.',
      descInternal:
               'It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout.',
      assetPath:
               'https://cdp-video-tst.s3.amazonaws.com/2020/01/commons.america.gov_ck5iguwcv01j40729bz48sijs/coffee.jpg',
      author: {
        id: 'dejkndejn',
        firstName: 'John',
        lastName: 'Doe',
        email: 'doej2@america.gov',
      },
      team: {
        id: 'wjbjb',
        name: 'GPA Design & Editorial',
        contentTypes: ['GRAPHIC'],
      },
      status: 'DRAFT',
      visibility: 'PUBLIC',
      images: [],
      categories: [],
      tags: [],
      __typename: 'GraphicProject',
    },
  ],
};
