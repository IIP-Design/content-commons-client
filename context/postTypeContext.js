import React from 'react';

const initialState = {
  error: false,
  list: [],
  loading: false,
};

export const PostTypeContext = React.createContext( initialState );

export const postTypeReducer = ( state, action ) => {
  const { payload } = action;

  switch ( action.type ) {
    case 'LOAD_POST_TYPES_PENDING':
      return {
        ...state,
        loading: true,
      };

    case 'LOAD_POST_TYPES_FAILED':
      return {
        ...state,
        error: true,
      };

    case 'LOAD_POST_TYPES_SUCCESS':
      return {
        ...state,
        error: false,
        loading: false,
        list: payload,
      };

    default:
      return state;
  }
};
