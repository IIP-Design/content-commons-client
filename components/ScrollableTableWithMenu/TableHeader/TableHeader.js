/**
 *
 * TableHeader
 *
 */
import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { Table } from 'semantic-ui-react';

import { DashboardContext } from 'context/dashboardContext';

import './TableHeader.scss';

const TableHeader = ( {
  tableHeaders,
  handleSort,
  displayActionsMenu
} ) => {
  const { state } = useContext( DashboardContext );

  return (
    <Table.Header>
      <Table.Row>
        { tableHeaders.map( header => (
          <Table.HeaderCell
            key={ header.name }
            sorted={ state.column === header.name ? state.direction : null }
            onClick={ handleSort( header.name ) }
            className={ displayActionsMenu ? 'displayActionsMenu' : '' }
          >
            { header.label }
          </Table.HeaderCell>
        ) ) }
      </Table.Row>
    </Table.Header>
  );
};

TableHeader.propTypes = {
  tableHeaders: PropTypes.array.isRequired,
  handleSort: PropTypes.func,
  displayActionsMenu: PropTypes.bool
};

export default TableHeader;
