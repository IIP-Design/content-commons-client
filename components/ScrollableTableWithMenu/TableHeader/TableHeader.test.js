import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { DashboardContext } from '../../../context/dashboardContext';
import TableHeader from './TableHeader';

const props = {
  tableHeaders: [
    { name: 'projectTitle', label: 'PROJECT TITLE' },
    { name: 'createdAt', label: 'CREATED' },
    { name: 'visibility', label: 'VISIBILITY' },
    { name: 'author', label: 'AUTHOR' }
  ],
  handleSort: jest.fn(),
  displayActionsMenu: false
};

jest.mock( 'react', () => ( {
  ...jest.requireActual( 'react' ),
  useContext: () => ( { state: { column: 'author', direction: 'descending' } } )
} ) );

const createTable = tableProps => (
  <DashboardContext.Provider>
    <table>
      <TableHeader { ...tableProps } />
    </table>
  </DashboardContext.Provider>
);

describe( '<TableHeader />', () => {
  it( 'renders without crashing', () => {
    const wrapper = mount( createTable( props ) );

    const header = wrapper.find( 'TableHeader' );

    expect( header.exists() ).toEqual( true );
    expect( toJSON( header ) ).toMatchSnapshot();
  } );

  it( 'renders the correct table header values', () => {
    const wrapper = mount( createTable( props ) );
    const th = wrapper.find( 'th' );

    th.forEach( ( t, i ) => {
      expect( t.text() ).toEqual( props.tableHeaders[i].label );
    } );
  } );

  it( 'renders the correct table header class value if displayActionsMenu', () => {
    const newProps = { ...props, displayActionsMenu: true };
    const wrapper = mount( createTable( newProps ) );

    const th = wrapper.find( 'th' );

    th.forEach( t => {
      expect( t.prop( 'className' ) ).toContain( 'displayActionsMenu' );
    } );
  } );

  it( 'renders the correct table header sorted and direction values if `column === header.name`', () => {
    const wrapper = mount( createTable( props ) );

    const cell = wrapper.find( 'TableHeaderCell' );

    cell.forEach( c => {
      if ( c.prop( 'children' ) === 'AUTHOR' ) {
        expect( c.prop( 'sorted' ) ).toEqual( 'descending' );
      } else {
        expect( c.prop( 'sorted' ) ).toEqual( null );
      }
    } );
  } );
} );
