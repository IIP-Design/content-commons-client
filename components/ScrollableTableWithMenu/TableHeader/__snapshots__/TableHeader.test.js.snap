// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableHeader /> renders without crashing 1`] = `
Array [
  <TableHeader
    displayActionsMenu={false}
    handleSort={
      [MockFunction] {
        "calls": Array [
          Array [
            "projectTitle",
          ],
          Array [
            "createdAt",
          ],
          Array [
            "visibility",
          ],
          Array [
            "author",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    tableHeaders={
      Array [
        Object {
          "label": "PROJECT TITLE",
          "name": "projectTitle",
        },
        Object {
          "label": "CREATED",
          "name": "createdAt",
        },
        Object {
          "label": "VISIBILITY",
          "name": "visibility",
        },
        Object {
          "label": "AUTHOR",
          "name": "author",
        },
      ]
    }
  >
    <TableHeader
      as="thead"
    >
      <thead
        className=""
      >
        <TableRow
          as="tr"
          cellAs="td"
        >
          <tr
            className=""
          >
            <TableHeaderCell
              as="th"
              className=""
              key="projectTitle"
              sorted={null}
            >
              <TableCell
                as="th"
                className=""
              >
                <th
                  className=""
                >
                  PROJECT TITLE
                </th>
              </TableCell>
            </TableHeaderCell>
            <TableHeaderCell
              as="th"
              className=""
              key="createdAt"
              sorted={null}
            >
              <TableCell
                as="th"
                className=""
              >
                <th
                  className=""
                >
                  CREATED
                </th>
              </TableCell>
            </TableHeaderCell>
            <TableHeaderCell
              as="th"
              className=""
              key="visibility"
              sorted={null}
            >
              <TableCell
                as="th"
                className=""
              >
                <th
                  className=""
                >
                  VISIBILITY
                </th>
              </TableCell>
            </TableHeaderCell>
            <TableHeaderCell
              as="th"
              className=""
              key="author"
              sorted="descending"
            >
              <TableCell
                as="th"
                className="descending sorted"
              >
                <th
                  className="descending sorted"
                >
                  AUTHOR
                </th>
              </TableCell>
            </TableHeaderCell>
          </tr>
        </TableRow>
      </thead>
    </TableHeader>
  </TableHeader>,
  <TableHeader
    as="thead"
  >
    <thead
      className=""
    >
      <TableRow
        as="tr"
        cellAs="td"
      >
        <tr
          className=""
        >
          <TableHeaderCell
            as="th"
            className=""
            key="projectTitle"
            sorted={null}
          >
            <TableCell
              as="th"
              className=""
            >
              <th
                className=""
              >
                PROJECT TITLE
              </th>
            </TableCell>
          </TableHeaderCell>
          <TableHeaderCell
            as="th"
            className=""
            key="createdAt"
            sorted={null}
          >
            <TableCell
              as="th"
              className=""
            >
              <th
                className=""
              >
                CREATED
              </th>
            </TableCell>
          </TableHeaderCell>
          <TableHeaderCell
            as="th"
            className=""
            key="visibility"
            sorted={null}
          >
            <TableCell
              as="th"
              className=""
            >
              <th
                className=""
              >
                VISIBILITY
              </th>
            </TableCell>
          </TableHeaderCell>
          <TableHeaderCell
            as="th"
            className=""
            key="author"
            sorted="descending"
          >
            <TableCell
              as="th"
              className="descending sorted"
            >
              <th
                className="descending sorted"
              >
                AUTHOR
              </th>
            </TableCell>
          </TableHeaderCell>
        </tr>
      </TableRow>
    </thead>
  </TableHeader>,
]
`;
